tools/querytee/proxy_endpoint_test.go

```diff
@@ -786,3 +786,74 @@ func (b *mockProxyBackend) ForwardRequest(_ context.Context, _ *http.Request, _
+			secondaryBackendRequestProportion:   1.0,
+			expectedPreferredOnlySelectionCount: runCount,
+		},
+		"single non-preferred backend, secondary request proportion 0.0": {
+			backends:                            []ProxyBackendInterface{newMockProxyBackend("preferred-backend", 0, false, nil)},
+			secondaryBackendRequestProportion:   0.0,
+			expectedPreferredOnlySelectionCount: runCount,
+		},
+		"single non-preferred backend, secondary request proportion 1.0": {
+			backends:                            []ProxyBackendInterface{newMockProxyBackend("preferred-backend", 0, false, nil)},
+			secondaryBackendRequestProportion:   1.0,
+			expectedPreferredOnlySelectionCount: runCount,
+		},
+		"multiple backends, secondary request proportion 0.0": {
+			backends:                            []ProxyBackendInterface{newMockProxyBackend("preferred-backend", 0, true, nil), newMockProxyBackend("non-preferred-backend", 0, false, nil)},
+			secondaryBackendRequestProportion:   0.0,
+			expectedPreferredOnlySelectionCount: runCount,
+		},
+		"multiple backends, secondary request proportion 0.2": {
+			backends:                            []ProxyBackendInterface{newMockProxyBackend("preferred-backend", 0, true, nil), newMockProxyBackend("non-preferred-backend", 0, false, nil)},
+			secondaryBackendRequestProportion:   0.2,
+			expectedPreferredOnlySelectionCount: 800,
+		},
+		"multiple backends, secondary request proportion 1.0": {
+			backends:                            []ProxyBackendInterface{newMockProxyBackend("preferred-backend", 0, true, nil), newMockProxyBackend("non-preferred-backend", 0, false, nil)},
+			secondaryBackendRequestProportion:   1.0,
+			expectedPreferredOnlySelectionCount: 0,
+		},
+	}
+
+	for name, testCase := range testCases {
+		t.Run(name, func(t *testing.T) {
+			proxyEndpoint := NewProxyEndpoint(testCase.backends, Route{}, nil, nil, nil, 0, testCase.secondaryBackendRequestProportion)
+			preferredOnlySelectionCount := 0
+
+			for i := 0; i < runCount; i++ {
+				backends := proxyEndpoint.selectBackends()
+				require.GreaterOrEqual(t, len(backends), 1)
+
+				if len(backends) == 1 {
+					preferredOnlySelectionCount++
+				}
+			}
+
+			if testCase.expectedPreferredOnlySelectionCount == 0 || testCase.expectedPreferredOnlySelectionCount == runCount {
+				// We expect to have selected only the preferred backend either every time or never.
+				require.Equal(t, testCase.expectedPreferredOnlySelectionCount, preferredOnlySelectionCount)
+			} else {
+				// We expect to have selected only the preferred backend just some of the time.
+				// Allow for some variation due to randomness.
+				require.InEpsilonf(t, testCase.expectedPreferredOnlySelectionCount, preferredOnlySelectionCount, 0.1, "expected to choose only the preferred backend %v times, but chose it %v times", testCase.expectedPreferredOnlySelectionCount, preferredOnlySelectionCount)
```

Comment:
I believe this will fail 1 in every 1000 test runs on average. I wonder if we should either increase the epsilon or number of runs. I think this test would still be accurate enough if we used an epsilon of 0.2. The tests are more about checking the edge cases in my opinion.

Reply:
Resolved in 9f8c7f1